### DISCUSSION TITLE
Tab key behaviour

### DISCUSSION THREAD

**Discussion by cpcallen** [RaspberryPiFoundation/blockly-keyboard-experimentation#63] [General]
Feedback from testers includes:

> People who are used to using keyboard navigation, expect to use TAB to get around and instinctively use that as the first key on arrival.

There are a few possible ways that Blockly could react to the tab key, including e.g.:

* Always cause focus to leave Blockly and move on to the next focusable element on the page.
* Move focus between toolbox / flyout / workspace but not within each area.  (See: #62)
* Move the cursor between and even within blocks.
  * This could be as fine-grained as doing a strict preorder traversal of the whole AST.
  * Alternatively the granularity could vary depending on how deep the cursor is—i.e., it could be basically a next-sibling button—but this will either mean that either the symmetry of tab and shift-tab is broken when you get bumped up a level a the end of a list of siblings, or (if tab won't bumped the cursor up a level) one would need to use a different keystroke to escape from Blockly.  (Perhaps the latter is OK, though, given that in this scenario one would have needed a different keystroke to get down to that level.)

Obviously the more fine-grained tab is within blockly the more difficult it will be to use tab to escape from Blockly to resume navigating the rest of the page.

**Comment by kmcnaught** [comment]
A few comments, 
- I think if the keyboard focus is more clearly bounded as per #49, this confusion becomes less problematic - if you reach for using Tab and discover the toolbox and code editor as 2 focusable elements, you can learn to navigate within them with arrow/other navigation, without breaking your expectations. Currently, however, there's a big unknown of "am I in the workspace?" which is confusing
- In the same vein, I would say the first goal is to have _page_ navigation with Tab to be predictable and standard, including working out where the code and toolbox belong on the page, with more scope to experiment _within_ code and toolbox
- If Tab is used within code then I'd expect Esc to get out completely (to tab to another part of the page, for example), although that _might_ cause some confusions if we are also using Esc to get out of a submenu like a colour palette or dropdown. Subject to testing, but I don't think this would be unreasonable, the convention of Escaping out of things is standard and clear. 
- If Tab is used within code/toolbox, there's a question of whether it's all in the main tab flow, or if you need to take an action to "enter" the subregions and then tab between things inside
- Anywhere that arrow key navigation is used, we need to consider how this compares to screen reader use of arrow keys. What you lose by overriding arrow key behaviour is the ability to step through a sentence word by word. Currently, I think that the scope of focus is small enough that this might not be a problem - it's akin to navigating a dropdown box / form with arrow keys, and hearing a full entry read out. If there's likely to be any more verbose content (like you see in the [micro:bit python editor](https://python.microbit.org/v/3/reference)'s reference pane) then this would be more of a concern, and in that scenario there would be more of an argument to use tab more and arrow keys less in this context. Are there likely to be any situations like this, with more verbose content? I am not aware of any in the blockly-based editors I'm aware of. 
- If the contents of the code/toolbox are in the main tab order, I'd expect there to (later) be skip-links and regions, to navigate more efficiently past them.

At the moment the approximate starting UX I'm picturing would be:
- tab between any webpage elements -> toolbox -> code region
- either hit Enter or press arrow keys to start navigating within a region
- if you exit a region, the arrow nav is no longer active (you could consider saving your location for later, but the focus wouldn't be live)
- including tab navigation within a region is an open question that we can explore once the basic page nav + arrow-nav-within-code have been tested and finetuned.

**Comment by BenHenning** [comment]
Between @cpcallen's and @kmcnaught's comments, I see two different approaches to consider here:
1. Include tab as part of the navigation experience _within_ code as a **primary** navigation tool (as is done with the rest of the page).
2. Only use tab for navigation outside of Blockly (but still be cooperative).

I think it makes sense to start with (2) and iterate on that, first, before considering (1). To that end, it seems that a reasonable first place to start for (2) would be:
- Tab navigation works at the page-element level with the only tabbable BLockly elements being the toolbox and code workspace.
- If the toolbox or code workspace is tabbed to, 'enter' must be pressed to select that element. At that point, tab navigation will stop working and Blockly navigation will take over until 'escape' is pressed.

I'd like thoughts on the second bit above since another approach we could take is keeping tab navigation enabled and using element selection (rather than 'enter'/'escape') immediately start navigation within the element.

Two factors to consider here are consistency with other elements (as @kmcnaught mentioned above) and context switching between the code workspace and the toolbox (such as when selecting blocks to insert). Needing to press 'escape' and then tab-navigation may introduce a bit more friction, but it will also reduce accidental navigation.

/cc @rachel-fenichel as well in case you have any thoughts.

**Reply by rachel-fenichel** [reply]
I'm slightly concerned about the context switching, but I think we could resolve that by making `T` in the workspace move focus to the toolbox without needing to go up and out. I think your proposal 2 is a good thiing to implement for upcoming testing.

**Reply by BenHenning** [reply]
FYI #71 was filed to track actual changes to try and address the original problem statement.
